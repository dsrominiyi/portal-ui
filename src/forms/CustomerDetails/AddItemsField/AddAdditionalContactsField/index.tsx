import React, { FunctionComponent } from 'react';

import AddItemsField from '..';
import AdditionalContactForm from '../../AdditionalContactForm';
import AdditionalContactsTable from './AdditionalContactsTable';

interface Props {
  additionalContacts?: AdditionalContact[];
  onChange(values: AdditionalContact[]): void;
}

export const AddAdditionalContactsField: FunctionComponent<Props> = ({
  additionalContacts,
  onChange
}: Props) => (
  <AddItemsField
    fieldText="Additional contacts (optional)"
    itemName="additional contact"
    form={AdditionalContactForm}
    table={AdditionalContactsTable}
    items={additionalContacts}
    onChange={onChange}
  />
);

AddAdditionalContactsField.displayName = 'AddAdditionalContactsField';
export default AddAdditionalContactsField;
