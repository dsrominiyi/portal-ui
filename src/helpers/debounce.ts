/**
 * Prevents an action from being called more that once every `wait` milliseconds
 * The initial triggering of the action will be delayed by `wait` milliseconds if `immediate` = false
 *
 * e.g. <button onClick={event => debounce(handleClick, 200)(event)} />
 */

let timeout: number | null;

const debounce = (action: (...args: any[]) => void, wait = 300, immediate = true) => (
  ...args: any[]
) => {
  const later = () => {
    timeout = null;
    if (!immediate) {
      action(...args);
    }
  };
  const callNow = immediate && !timeout;

  if (timeout) {
    clearTimeout(timeout);
  }

  timeout = (setTimeout as Window['setTimeout'])(later, wait);

  if (callNow) {
    action(...args);
  }
};

export default debounce;
