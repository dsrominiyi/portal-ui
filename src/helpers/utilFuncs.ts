import { utc as moment } from 'moment';

type Primitive = string | number | boolean;

export const getKeyByValue = (object: { [key: string]: Primitive }, value?: Primitive) => {
  return Object.keys(object).find(key => object[key] === value);
};

export const stripNonNumeric = (value: string) => value.replace(/\D/g, '');

export const dateFieldValuesToDateString = (values: DateFieldValues) => {
  const { yyyy, mm, dd } = values;
  if (!yyyy && !mm && !dd) {
    return '';
  }
  if (yyyy.length !== 4 || mm.length !== 2 || dd.length !== 2) {
    return 'INVALID DATE';
  }

  return moment(`${yyyy}-${mm}-${dd}`, 'YYYY-MM-DD').toISOString() || 'INVALID DATE';
};

export const dateStringToDateFieldValues = (date: string): DateFieldValues => {
  const dd = moment(date, 'YYYY-MM-DD').format('DD');
  const mm = moment(date, 'YYYY-MM-DD').format('MM');
  const yyyy = moment(date, 'YYYY-MM-DD').format('YYYY');

  return { dd, mm, yyyy };
};

export const capitalise = (text: string) => {
  const words = text.split(/\s/g);
  return words.map(word => `${word.charAt(0).toUpperCase()}${word.slice(1)}`).join(' ');
};

export const kebabCase = (text: string) => {
  const words = text.split(/\s/g);
  return words.map(word => word.toLowerCase()).join('-');
};
