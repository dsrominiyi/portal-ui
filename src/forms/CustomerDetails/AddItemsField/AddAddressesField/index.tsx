import React, { FunctionComponent } from 'react';

import AddItemsField from '..';
import AddressForm from '../../AddressForm';
import AddressesTable from './AddressesTable';

interface Props {
  customerType: CustomerType;
  addresses?: Address[];
  hasError?: boolean;
  onChange(values: Address[]): void;
}

export const AddAddressesField: FunctionComponent<Props> = ({
  customerType,
  addresses,
  hasError,
  onChange
}: Props) => (
  <AddItemsField
    itemName="address"
    form={AddressForm}
    table={AddressesTable}
    items={addresses}
    formConfig={{ customerType }}
    hasError={hasError}
    errorText="Please add at least one address"
    onChange={onChange}
  />
);

AddAddressesField.displayName = 'AddAddressesField';
export default AddAddressesField;
