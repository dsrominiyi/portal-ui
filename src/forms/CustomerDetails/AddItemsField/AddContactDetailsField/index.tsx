import React, { FunctionComponent } from 'react';

import AddItemsField from '..';
import ContactDetailForm from '../../ContactDetailForm';
import ContactDetailsTable from './ContactDetailsTable';

interface Props {
  contactDetails?: Contact[];
  hasError?: boolean;
  onChange(values: Contact[]): void;
}

export const AddContactDetailsField: FunctionComponent<Props> = ({
  contactDetails,
  hasError,
  onChange
}: Props) => (
  <AddItemsField
    fieldText="Contact details"
    itemName="contact"
    form={ContactDetailForm}
    table={ContactDetailsTable}
    items={contactDetails}
    hasError={hasError}
    errorText="Please add at least one method of contact"
    onChange={onChange}
  />
);

AddContactDetailsField.displayName = 'AddContactDetailsField';
export default AddContactDetailsField;
