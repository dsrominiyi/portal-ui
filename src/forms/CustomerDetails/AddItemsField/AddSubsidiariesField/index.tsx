import React, { FunctionComponent } from 'react';

import AddItemsField from '..';
import SubsidiaryForm from '../../SubsidiaryForm';
import SubsidiariesTable from './SubsidiariesTable';

interface Props {
  subsidiaries?: Company[];
  onChange(values: Company[]): void;
}

export const AddSubsidiariesField: FunctionComponent<Props> = ({
  subsidiaries,
  onChange
}: Props) => (
  <AddItemsField
    fieldText="Subsidiaries (optional)"
    itemName="subsidiary"
    form={SubsidiaryForm}
    table={SubsidiariesTable}
    items={subsidiaries}
    onChange={onChange}
  />
);

AddSubsidiariesField.displayName = 'AddSubsidiariesField';
export default AddSubsidiariesField;
