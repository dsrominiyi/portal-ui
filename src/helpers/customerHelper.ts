import { dateFieldValuesToDateString, dateStringToDateFieldValues } from './utilFuncs';
import { predefinedRules, ValidationRules } from './formValidator';

export const companyFormValidationRules: ValidationRules = {
  companyName: { ...predefinedRules.entityName, fieldLabel: 'company name', required: true },
  tradingName: { ...predefinedRules.entityName, fieldLabel: 'trading name' },
  typeStatus: { fieldLabel: 'company legal status', invalidValues: ['Other'], required: true },
  businessClass: { fieldLabel: 'business class', required: true },
  tradingSinceDate: { fieldLabel: 'trading since', isDate: true },
  parentCompany: { fieldLabel: 'is this company a subsidiary', required: true },
  parentCompanyName: { ...predefinedRules.entityName, fieldLabel: 'parent company name' },
  registeredCoNumber: predefinedRules.registeredCoNumber,
  contactDetails: { fieldLabel: 'contact details', required: true },
  addresses: { fieldLabel: 'addresses', required: true }
};

export const isIndividualCustomer = (customer: Customer): customer is Individual => {
  return typeof (customer as Individual).name !== 'undefined';
};

export const getIndividualFullName = ({ title, forename, middleName, surname }: IndividualName) => {
  return `${title} ${forename}${middleName && ` ${middleName}`} ${surname}`;
};

export const buildCompanyFromFormData = (data: NewCompanyData): Company => ({
  ...data,
  parentCompany: data.parentCompany || false,
  tradingSinceDate: dateFieldValuesToDateString(data.tradingSinceDate)
});

export const buildFormDataFromCompany = (data: Company): NewCompanyData => ({
  ...data,
  tradingSinceDate: data.tradingSinceDate
    ? dateStringToDateFieldValues(data.tradingSinceDate)
    : { dd: '', mm: '', yyyy: '' }
});
