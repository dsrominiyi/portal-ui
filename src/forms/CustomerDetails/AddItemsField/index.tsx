import React, { FunctionComponent, useState, useEffect, Fragment } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import Button from '../../../components/Button';
import Modal from '../../../components/Modal';

import { capitalise, kebabCase } from '../../../helpers/utilFuncs';

import formStyles from '../../styles.scss';
import modalStyles from '../../../components/Modal/styles.scss';

interface Props {
  fieldText?: string;
  itemName: string;
  form: FunctionComponent<FormProps>;
  table: FunctionComponent<ItemTableProps>;
  items?: unknown[];
  formConfig?: { [key: string]: unknown };
  hasError?: boolean;
  errorText?: string;
  onChange(values: unknown[]): void;
}

interface State {
  items: unknown[];
  newItem?: unknown;
  isModalVisible: boolean;
  editIndex?: number;
  isFormValid: boolean;
  submissionAttempted: boolean;
}

type UseStateArray = [State, React.Dispatch<React.SetStateAction<State>>];

export const AddItemsField: FunctionComponent<Props> = ({
  fieldText,
  itemName,
  form: Form,
  table: Table,
  items: initialItems,
  formConfig,
  hasError,
  errorText,
  onChange
}: Props) => {
  const [state, setState]: UseStateArray = useState({
    items: initialItems || [],
    newItem: undefined,
    isModalVisible: false,
    editIndex: undefined,
    isFormValid: false,
    submissionAttempted: false
  } as State);

  useEffect(() => {
    onChange(state.items);
  }, [state.items]);

  const testIdItemName = kebabCase(itemName);
  const itemDisplayName = capitalise(itemName);

  const dismissModal = (updatedItems?: unknown[]) => {
    setState({
      ...state,
      isModalVisible: false,
      submissionAttempted: false,
      editIndex: undefined,
      items: updatedItems || state.items
    });
  };

  const saveItem = () => {
    const { items, newItem, editIndex } = state;

    if (newItem) {
      const updatedItems =
        typeof editIndex === 'number'
          ? [...items.slice(0, editIndex), newItem, ...items.slice(editIndex + 1)]
          : [...items, newItem];

      dismissModal(updatedItems);
    }
  };

  const renderModal = () => {
    const { items, isModalVisible, editIndex, isFormValid: canSubmit, submissionAttempted } = state;
    const editMode = typeof editIndex === 'number';

    const dismiss = () => dismissModal();
    const save = () => saveItem();

    return (
      <Modal
        isVisible={isModalVisible}
        body={[
          <Form
            key="0"
            additionalClasses={[formStyles.modalForm]}
            initialValues={editMode ? items[editIndex!] : undefined}
            submissionAttempted={submissionAttempted}
            config={formConfig}
            onChange={(values, isFormValid) => {
              setState({ ...state, newItem: values, isFormValid });
            }}
            onSubmit={save}
          />
        ]}
        header={`${editMode ? 'Edit' : 'Add'} ${itemDisplayName}`}
        footer={
          <div className={modalStyles.footerActions}>
            <Button
              label="Cancel"
              onClick={dismiss}
              attributes={{ 'test-id': `dismiss-${testIdItemName}-modal` }}
            />
            <Button
              label={editMode ? 'Update' : 'Add'}
              onClick={() => {
                setState({ ...state, submissionAttempted: true });
                return canSubmit && save();
              }}
              status={Status.Success}
              attributes={{ 'test-id': `save-${testIdItemName}` }}
            />
          </div>
        }
        hideModal={dismiss}
      />
    );
  };

  const renderTable = () => {
    const { items } = state;
    const onEdit = (editIndex: number) => {
      setState({
        ...state,
        isModalVisible: true,
        editIndex
      });
    };
    const onDelete = (deleteIndex: number) => {
      const updatedItems = [...items];
      updatedItems.splice(deleteIndex, 1);
      setState({
        ...state,
        items: updatedItems
      });
    };

    return items.length > 0 && <Table key="1" items={items} onEdit={onEdit} onDelete={onDelete} />;
  };

  return (
    <Fragment>
      <QuestionBox
        text={fieldText}
        hasError={hasError}
        errorText={errorText}
        inputFields={[
          <Button
            key="0"
            label={`Add ${itemDisplayName}`}
            onClick={() => setState({ ...state, isModalVisible: true })}
            status={Status.Success}
            attributes={{ 'test-id': `add-${testIdItemName}` }}
          />,
          renderTable()
        ]}
      />
      {renderModal()}
    </Fragment>
  );
};

AddItemsField.displayName = 'AddItemsField';
export default AddItemsField;
