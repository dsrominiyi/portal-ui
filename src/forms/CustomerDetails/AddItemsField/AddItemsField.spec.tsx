import React, { FunctionComponent } from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

import AddItemsField from '.';

import { capitalise } from '../../../helpers/utilFuncs';

jest.mock('../../../helpers/domHelper');

let component: ShallowWrapper | ReactWrapper;
const fieldText = 'Add items';
const itemName = 'new item';
const item = {
  name: 'item'
};
const item2 = {
  name: 'item2'
};
const form: FunctionComponent<FormProps> = () => <div />;
const table: FunctionComponent<ItemTableProps> = () => <div />;
const onChange = jest.fn();

const getAddButton = () => (component.find('QuestionBox').props() as any).inputFields[0];
const getTable = () => (component.find('QuestionBox').props() as any).inputFields[1];
const getForm = () => (component.find('Modal').props() as any).body[0];

const clickAddButton = () => {
  act(() => {
    getAddButton().props.onClick();
  });
  component.update();
};

const saveItem = (itemToSave?: unknown) => {
  act(() => {
    getForm().props.onChange(itemToSave, true);
  });
  component.update();
  act(() => {
    getForm().props.onSubmit();
  });
  component.update();
};

const setEditMode = (editIndex: number) => {
  act(() => {
    getTable().props.onEdit(editIndex);
  });
  component.update();
};

const deleteItem = (deleteIndex: number) => {
  act(() => {
    getTable().props.onDelete(deleteIndex);
  });
  component.update();
};

const render = (items?: unknown[]) => (
  <AddItemsField
    fieldText={fieldText}
    itemName={itemName}
    form={form}
    table={table}
    items={items}
    onChange={onChange}
  />
);

describe('AddItemsField', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    component = mount(render());
  });

  test('AddItemsField matches snapshot', () => {
    component = shallow(render());
    expect(component).toMatchSnapshot();
  });

  test('AddItemsField with initial items matches snapshot', () => {
    component = shallow(render([item, item2]));
    expect(component).toMatchSnapshot();
  });

  test('clicking the add button reveals the modal', () => {
    clickAddButton();
    const modal = component.find('Modal');

    expect((modal.props() as any).isVisible).toBe(true);
    expect((modal.props() as any).header).toBe(`Add ${capitalise(itemName)}`);
  });

  test('clicking "cancel" in the modal footer dismisses it without saving', () => {
    clickAddButton();
    const cancelButton = shallow((component.find('Modal').props() as any).footer)
      .find('Button')
      .at(0);

    act(() => {
      cancelButton.simulate('click');
    });
    component.update();
    const modal = component.find('Modal');

    expect((modal.props() as any).isVisible).toBe(false);
  });

  test('clicking "Add" in the modal sets the submissionAttempted state value to true', () => {
    clickAddButton();
    const modalSaveButton = shallow((component.find('Modal').props() as any).footer)
      .find('Button')
      .at(1);

    act(() => {
      modalSaveButton.simulate('click');
    });
    component.update();

    expect(getForm().props.submissionAttempted).toBe(true);
  });

  test('clicking "Add" in the modal does not save the item if the form is invalid', () => {
    clickAddButton();
    const modalSaveButton = shallow((component.find('Modal').props() as any).footer)
      .find('Button')
      .at(1);

    act(() => {
      modalSaveButton.simulate('click');
    });
    component.update();

    expect(onChange.mock.calls[1]).toBeUndefined();
  });

  test('saving the form calls onChange with the updated items', () => {
    saveItem(item);

    expect(onChange).toHaveBeenCalledWith([item]);
  });

  test('clicking "Add" in the modal saves the item if the form is valid', () => {
    clickAddButton();
    act(() => {
      getForm().props.onChange(item, true);
    });
    component.update();
    const modalSaveButton = shallow((component.find('Modal').props() as any).footer)
      .find('Button')
      .at(1);

    act(() => {
      modalSaveButton.simulate('click');
    });
    component.update();

    expect(onChange).toHaveBeenCalledWith([item]);
  });

  test('the table is rendered when there are saved items', () => {
    expect(getTable()).toBe(false);

    saveItem(item);

    const items = onChange.mock.calls[1][0];
    expect(getTable().props.items).toEqual(items);
  });

  test('clicking the edit button of a row in the table opens an edit modal', () => {
    saveItem(item);
    setEditMode(0);
    const modal = component.find('Modal');

    expect((modal.props() as any).isVisible).toBe(true);
    expect((modal.props() as any).header).toBe(`Edit ${capitalise(itemName)}`);
    expect(getForm().props.initialValues).toEqual(item);
  });

  test('saving an edited item updates the correct index under items', () => {
    const editedContact = {
      ...item,
      relationship: AdditionalContactRelationship.Parent
    };

    saveItem(item);
    saveItem(item2);
    setEditMode(0);
    saveItem(editedContact);
    const items = onChange.mock.calls[3][0];

    expect(items).toEqual([editedContact, item2]);
  });

  test('attempting to save an item takes no action if state.newItem is undefined', () => {
    saveItem();

    expect(onChange.mock.calls[1]).toBeUndefined();
  });

  test('clicking the delete button of a row in the table removes it from state', () => {
    saveItem(item);
    saveItem(item2);
    deleteItem(0);
    const items = onChange.mock.calls[3][0];

    expect(items).toEqual([item2]);
  });
});
