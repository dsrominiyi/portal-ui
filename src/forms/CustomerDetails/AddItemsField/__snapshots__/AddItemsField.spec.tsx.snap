// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItemsField AddItemsField matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    inputFields={
      Array [
        <Button
          attributes={
            Object {
              "test-id": "add-new-item",
            }
          }
          label="Add New Item"
          onClick={[Function]}
          status="Success"
        />,
        false,
      ]
    }
    text="Add items"
  />
  <Modal
    body={
      Array [
        <form
          additionalClasses={
            Array [
              "modalForm",
            ]
          }
          onChange={[Function]}
          onSubmit={[Function]}
          submissionAttempted={false}
        />,
      ]
    }
    footer={
      <div
        className="footerActions"
      >
        <Button
          attributes={
            Object {
              "test-id": "dismiss-new-item-modal",
            }
          }
          label="Cancel"
          onClick={[Function]}
        />
        <Button
          attributes={
            Object {
              "test-id": "save-new-item",
            }
          }
          label="Add"
          onClick={[Function]}
          status="Success"
        />
      </div>
    }
    header="Add New Item"
    hideModal={[Function]}
    isVisible={false}
  />
</Fragment>
`;

exports[`AddItemsField AddItemsField with initial items matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    inputFields={
      Array [
        <Button
          attributes={
            Object {
              "test-id": "add-new-item",
            }
          }
          label="Add New Item"
          onClick={[Function]}
          status="Success"
        />,
        <table
          items={
            Array [
              Object {
                "name": "item",
              },
              Object {
                "name": "item2",
              },
            ]
          }
          onDelete={[Function]}
          onEdit={[Function]}
        />,
      ]
    }
    text="Add items"
  />
  <Modal
    body={
      Array [
        <form
          additionalClasses={
            Array [
              "modalForm",
            ]
          }
          onChange={[Function]}
          onSubmit={[Function]}
          submissionAttempted={false}
        />,
      ]
    }
    footer={
      <div
        className="footerActions"
      >
        <Button
          attributes={
            Object {
              "test-id": "dismiss-new-item-modal",
            }
          }
          label="Cancel"
          onClick={[Function]}
        />
        <Button
          attributes={
            Object {
              "test-id": "save-new-item",
            }
          }
          label="Add"
          onClick={[Function]}
          status="Success"
        />
      </div>
    }
    header="Add New Item"
    hideModal={[Function]}
    isVisible={false}
  />
</Fragment>
`;
